Audit log blog post

gunnarmorling opened the pull request:
Blog post accompanying the [audit log demo](https://github.com/debezium/debezium-examples/tree/master/auditlog). Feedback welcome!

hpgrahsl commented on blog/2019-09-12-audit-logs-with-change-data-capture-and-kafka-streams.adoc:
```diff
@@ -0,0 +1,647 @@
+
+Library-based solutions also oftentimes cannot provide metadata in case of bulk updates and deletes (`UPDATE mytable WHERE ...`),
+as the listeners used to hook them into the persistence framework are not aware of all the affected records.
+Also, by definition, any approach based on an application-side library is prone to miss any data changes done directly in the database, e.g. when running a data load or bypassing the application during an emergency data patch.
```
Also here the same holds true for the proposed solution. Since direct changes won't trigger application level meta data insertion which is done in the example code. The only difference is that you can e.g. "block the enrichments" in the stream processor but still there is no such data available under such circumstances.

gunnarmorling replied:
Yes, that's true as far as the metadata information is concerned. The actual change events of course would be captured from the log also for a bulk update/delete. I'll see how I can rephrase that; I was unsure about that myself, but thought I'd get away with it. Seems that not :)

gunnarmorling replied:
I hope it's clearer now with the rewording of this section.

hpgrahsl commented on blog/2019-09-12-audit-logs-with-change-data-capture-and-kafka-streams.adoc:
```diff
@@ -0,0 +1,647 @@
+Also, by definition, any approach based on an application-side library is prone to miss any data changes done directly in the database, e.g. when running a data load or bypassing the application during an emergency data patch.
+
+These problems don't exist when leveraging the transaction log as the source for an audit trail and using change data capture for retrieving the change information and sending it into a message broker or log such as Apache Kafka.
+Running asynchronously, the CDC process can extract the change data without impacting OLTP transactions, not missing any event, no matter whether what its origin is and whether it was caused by a bulk operation.
```
...,no matter ~whether~ what its origin and whether ...

hpgrahsl commented on blog/2019-09-12-audit-logs-with-change-data-capture-and-kafka-streams.adoc:
```diff
@@ -0,0 +1,647 @@
+}
+----
+
+Not too exciting, if you've ever built REST services with JAX-RS before:
```
I'd rephrase that somehow even though you are right about the "excitment" this might give readers :-)

gunnarmorling replied:
Yes, good point. It didn't feel quite right to me, too, but I couldn't quite say what it is. Will reword.

hpgrahsl commented on blog/2019-09-12-audit-logs-with-change-data-capture-and-kafka-streams.adoc:
```diff
@@ -0,0 +1,647 @@
+| 11   | Delicious!    | Banana  |
+| 10   | tasty         | Pear    |
++------+---------------+---------+
+SELECT 3
```
I would suggest to remove this info about rows + timing. WDYT?

hpgrahsl commented on blog/2019-09-12-audit-logs-with-change-data-capture-and-kafka-streams.adoc:
```diff
@@ -0,0 +1,647 @@
+it's time to set up the link:/documentation/reference/0.10/connectors/postgresql.html[Debezium Postgres connector] and stream the data changes from the `vegetable` and `transaction_context_data` tables into corresponding Kafka topics.
+Again refer to the example README file for the details of https://github.com/debezium/debezium-examples/tree/master/auditlog#deploy-the-debezium-postgres-connector[deploying the connector].
+
+The `dbserver1.inventory.vegetable` topic should contain contain change events for created, updated and deleted vegetable records, whereas the `dbserver1.inventory.transaction_context_data` topic should only contain create messages for each inserted metadata record.
```
"contain" 2x

hpgrahsl commented on blog/2019-09-12-audit-logs-with-change-data-capture-and-kafka-streams.adoc:
```diff
@@ -0,0 +1,647 @@
+++++
+
+This kind of message transformation is a perfect use case for https://kafka.apache.org/documentation/streams/[Kafka Streams],
+a Java API for implementing stream processing applications on top of on Kafka topics,
```
...on top of ~on~ Kafka topics,

hpgrahsl commented on blog/2019-09-12-audit-logs-with-change-data-capture-and-kafka-streams.adoc:
```diff
@@ -0,0 +1,647 @@
+This means that all messages for the same vegetable record will have the same key and thus will go into the same partition of the vegetables Kafka topic.
+This in turn guarantees that a consumer of these events sees all the messages pertaining to the same vegetable record in the exact same order as they were created.
+
+Now, in order to join the two streams, the message key must be the same on both sides.
```
You might add just 1 sentence that a stream-stream join would also need a window for a length which is not easy to define in general.

gunnarmorling replied:
Done.

hpgrahsl commented on blog/2019-09-12-audit-logs-with-change-data-capture-and-kafka-streams.adoc:
```diff
@@ -0,0 +1,647 @@
+* define a `GlobalKTable` based on the transaction metadata topic and
+* define a `KStream` based on the vegetables topic; on this stream, any incoming tombstone markers are filtered, the reasoning being that the retention policy for an audit trail topic typically should be time-based than based on log compaction; similarly, snapshot events are filtered, assuming they are not relevant for an audit trail and there wouldn't be any corresponding metadata provided by th ...
++
+Any other messages are enriched with the corresponding transaction metadata via a custom `Transformer` (see below) and finally are written to an output topic
```
and finally ~are~ written

gunnarmorling replied:
Hum, I think this was correct?

hpgrahsl commented on blog/2019-09-12-audit-logs-with-change-data-capture-and-kafka-streams.adoc:
```diff
@@ -0,0 +1,647 @@
+
+The basic logic of the transformer is this:
+
+* When a vegetables change event arrives, look up the corresponding metadata in the statestore of the 
```
statestore -> state store

hpgrahsl commented on blog/2019-09-12-audit-logs-with-change-data-capture-and-kafka-streams.adoc:
```diff
@@ -0,0 +1,647 @@
+* Before actually getting to the incoming event, all buffered events are processed; only if all could be enriched, the incoming event will be processed, too
+* In order to emit buffered events also if no new change event is coming in, the `punctuate()` method is used to periodically process the buffer
+
+The key piece is the buffer for unprocesseable change events.
```
unprocess-e-able

hpgrahsl commented on blog/2019-09-12-audit-logs-with-change-data-capture-and-kafka-streams.adoc:
```diff
@@ -0,0 +1,647 @@
+* In order to emit buffered events also if no new change event is coming in, the `punctuate()` method is used to periodically process the buffer
+
+The key piece is the buffer for unprocesseable change events.
+To maintain the order of events, the buffer must be processed in order of insertion,
```
FIFO might help to add to the sentence somewhere?

hpgrahsl commented on blog/2019-09-12-audit-logs-with-change-data-capture-and-kafka-streams.adoc:
```diff
@@ -0,0 +1,647 @@
+The implemention is based on the assumption that change events are serialized as JSON,
+but of course it could be done equally well using other formats such as Avro or Protocol Buffers.
+
+This is a bit of code, but hopefully its decomposition into multiple smaller methods makes it comprehensible:
```
Since you describe only the most relevant parts below the code block maybe there is not need to provide all code lines here but instead reduce it to the code blocks that are also discussed in the text and link to the source file. WDYT?

gunnarmorling replied:
Yes, I was wondering that, too. But I've added AsciiDoc callouts now, which I think help a lot with keeping this comprehensible (hard to see here on GitHub, but they'll show up in the rendered output on the website). So I think it's ok as is. I've already removed the `BufferOffsets` class and just point to it on GitHub.

gunnarmorling replied:
Here's an image of the rendered callouts; they're referenced in text below the listing. <img width="859" alt="Screen Shot 2019-09-29 at 22 14 43" src="https://user-images.githubusercontent.com/28612/65838747-ae97be80-e306-11e9-9b45-b7670e31a7e3.png">

hpgrahsl commented on blog/2019-09-12-audit-logs-with-change-data-capture-and-kafka-streams.adoc:
```diff
@@ -0,0 +1,647 @@
+e.g. to examine specific earlier versions of the data.
+To do so, the enriched change data events typically would be stored in a queryable database.
+Unlike a basic data replication pipeline, not only the latest version of each record would be stored in the database in that case, but all the versions, i.e. the primary keys typically would be amended with the transaction id of each change.
+This would all to select single data records or even joins of multiple tables to get the data valid as per a given transaction id.
```
all -> allow

hpgrahsl commented on blog/2019-09-12-audit-logs-with-change-data-capture-and-kafka-streams.adoc:
```diff
@@ -0,0 +1,647 @@
+To do so, the enriched change data events typically would be stored in a queryable database.
+Unlike a basic data replication pipeline, not only the latest version of each record would be stored in the database in that case, but all the versions, i.e. the primary keys typically would be amended with the transaction id of each change.
+This would all to select single data records or even joins of multiple tables to get the data valid as per a given transaction id.
+How this could be implemented in detail, may be discussed in a future post.
```
no comma needed I guess?

gunnarmorling commented on blog/2019-09-12-audit-logs-with-change-data-capture-and-kafka-streams.adoc:
```diff
@@ -0,0 +1,647 @@
+This comes with some disadvantages, though:
+
+* metadata columns and triggers add bloat to the model in the database and the corresponding application domain model
+* writing records in history tables as part of OLTP transactions increases the number of executed statements within the transaction and thus may cause longer response times of the application
```
The difference is in the number of records; indeed some kind of metadata record will be written in both cases. But a library like Envers or a triggers based approach will *also* write the entries in the history table: for each update and delete there'd be another insert in the history table. Whereas this is not the case for the proposed CDC-based solution, where that information is obtained from the log.

gunnarmorling replied:
I've added a sentence to make it clearer:

 >writing records in history tables as part of OLTP transactions increases the number of executed statements within the transaction (for each update or delete, also an insert must be written into the corresponding history table) and thus may cause longer response times of the application

Does that work better?

gunnarmorling replied:
I've largely restructured this section. It should be much clearer now.

gunnarmorling commented:
Thanks a lot for the thorough proof-reading, @hpgrahsl! I've pushed one more commit addressing most your remarks and resolved the corresponding discussions.

jbfletch commented:
Still reading but one comment is that all your Farmers in the transactions table are men as a lady farmer it would be nice to see a Farmer Margaret 🙂

jpechane commented on blog/2019-09-12-audit-logs-with-change-data-capture-and-kafka-streams.adoc:
```diff
@@ -0,0 +1,656 @@
+----
+
+[source,sql]
+----
```
Minor nit pick, why in the example above in the trasactions table the usernames are with space and below without?

gunnarmorling replied:
Good catch; Updating the image.

jpechane commented on blog/2019-09-12-audit-logs-with-change-data-capture-and-kafka-streams.adoc:
```diff
@@ -0,0 +1,656 @@
+a Java API for implementing stream processing applications on top of Kafka topics,
+providing operators that let you filter, transform, aggregate and join Kafka messages.
+
+As foundation for our stream processing application we're going to use https://quarkus.io/[Quarkus],
```
Is not runtime environment better word than foundaiton in this context?

jpechane commented on blog/2019-09-12-audit-logs-with-change-data-capture-and-kafka-streams.adoc:
```diff
@@ -0,0 +1,656 @@
+transaction topic's `GlobalKTable`, using the transaction id from the `source` block of the change event as the key;
+if the metadata could be found, add the metadata to change event (under the `audit` field) and return that enriched event
+<2> If the metadata could not be found, add the incoming event into the buffer of change events and return
+<3> Before actually getting to the incoming event, all buffered events are processed; only if all could be enriched, the incoming event will be processed, too
```
You should explain why

gunnarmorling commented:
> Still reading but one comment is that all your Farmers in the transactions table are men as a lady farmer it would be nice to see a Farmer Margaret 

@jbfletch, uff, absolutely, will update. Sorry Barry, Margaret it is :)

gunnarmorling commented on blog/2019-09-12-audit-logs-with-change-data-capture-and-kafka-streams.adoc:
```diff
@@ -0,0 +1,656 @@
+= Building Audit Logs with Change Data Capture and Stream Processing
```
Yes, will do. Was only waiting with it until the end, as I didn't want to mess up with the review comments.

gunnarmorling commented on blog/2019-09-12-audit-logs-with-change-data-capture-and-kafka-streams.adoc:
```diff
@@ -0,0 +1,656 @@
+* metadata columns and triggers add bloat to the model in the database and the corresponding application domain model
+* writing records in history tables as part of OLTP transactions increases the number of executed statements within the transaction (for each update or delete, also an insert must be written into the corresponding history table) and thus may cause longer response times of the application
+
+Library-based solutions also oftentimes cannot provide metadata in case of bulk updates and deletes (`UPDATE mytable WHERE ...`),
```
Clarified that the TX log will contain one entry for each record updated/deleted in a bulk operation.

gunnarmorling commented:
Pushed one more commit to address Jiri's remarks and add Margaret :)

gunnarmorling commented:
Thanks a lot for all the feedback, everyone! I think I've incorporated all the suggestion and am quite happy about this piece. I'll push it to the website later today. Thanks again!

gunnarmorling commented:
And it's live: https://twitter.com/gunnarmorling/status/1179062410134638594 :)
